[wip] rangefeed: improve memory accounting for event queue

wenyihu6 opened the pull request:
We can use the processor's resolved timestamp to predict if a checkpoint event will happen. But calling p.rts.ConsumeLogicalOp for memory accounting can change internal field unresolvedIntentQueue. To resolve this, we could make a deep copy of rts, but this is on a hot path and could introduce significant overhead. 

Related: https://github.com/cockroachdb/cockroach/issues/73616
Resolves: https://github.com/cockroachdb/cockroach/issues/114190

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120001)
<!-- Reviewable:end -->

wenyihu6 committed:
a

wenyihu6 committed:
refactor

wenyihu6 committed:
a

wenyihu6 committed:
a

erikgrinaker reviewed (commented):
Nice, the overall approach here makes sense. Leaving some minor comments.

> Reviewed 1 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120001#-:-NsmjE-YBXAc7FEyHQS_:b-t1oubg)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/rangefeed/event_size.go` line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120001#-Nsn1G2t-PaPjjSj1fZv:-Nsn1G2t-PaPjjSj1fZw:b5n4lne) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a428103984f0ff4b88637efed80ce1af83b624/pkg/kv/kvserver/rangefeed/event_size.go#L91)):*
> ```Go
> 
> func checkpointRangeFeedEvent(span roachpb.RSpan, ts resolvedTimestamp) kvpb.RangeFeedEvent {
> 	var event kvpb.RangeFeedEvent
> ```

Will these cause memory allocations? We should avoid those on the hot path (we do a bunch of them already, but should try not to add any more). You can find out by writing a small benchmark using `dev bench --bench-mem`, or using `scripts/goescape` to see if any of these escape to the heap.
___
*[`pkg/kv/kvserver/rangefeed/event_size.go` line 110 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120001#-NsmzndPCajUO0jtogZu:-NsmzndPCajUO0jtogZv:b-cf5nh0) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a428103984f0ff4b88637efed80ce1af83b624/pkg/kv/kvserver/rangefeed/event_size.go#L110)):*
> ```Go
> 
> func (ct ctEvent) futureMemUsage(span roachpb.RSpan, rts resolvedTimestamp) int64 {
> 	if rts.ForwardClosedTS(context.Background(), ct.Timestamp) {
> ```

We shouldn't call `ForwardClosedTS` here, nor `rts.Init()` for `initRTS`. These will actually advance the rangefeed's resolved timestamp, which will cause incorrect behavior when a later call e.g. during `consumeEvent` sees `ForwardClosedTS` return `false` because the resolved timestamp was already advanced when computing the event size.
___
*[`pkg/kv/kvserver/rangefeed/event_size.go` line 196 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120001#-Nsn1uhiEojSCPgH1bJB:-Nsn1uhiEojSCPgH1bJC:b-yllpk0) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a428103984f0ff4b88637efed80ce1af83b624/pkg/kv/kvserver/rangefeed/event_size.go#L196)):*
> ```Go
> 	switch re.GetValue().(type) {
> 	case *kvpb.RangeFeedValue:
> 		memUsage += rangefeedValueOverhead + int64(re.Size())
> ```

`Size()` won't really give an accurate representation of the memory size, as it computes the Protobuf-encoded wire size (which uses stuff like varints to compress integers, and omits fields with zero values).

I wonder if it would be sufficient here to simply compute the `event` memory usage, and assume that the Protobuf events have a size that's roughly equal to it. In other words, just use `memUsage()` and drop `futureMemUsage()`. How large is the diff usually?
___
*[`pkg/kv/kvserver/rangefeed/scheduled_processor.go` line 703 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120001#-Nsn2mC181TPNz3cs_L3:-Nsn2mC181TPNz3cs_L4:bn7gfpa) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a428103984f0ff4b88637efed80ce1af83b624/pkg/kv/kvserver/rangefeed/scheduled_processor.go#L703)):*
> ```Go
> 		if p.rts.ConsumeLogicalOp(ctx, op) {
> 			// May publish checkpoint we are over accounting here -> we should have a separate token for publishcheckpoint maybe
> 			p.publishCheckpoint(ctx, alloc)
> ```

I think this is fine, but we could also consider just passing a `nil` alloc here and ignoring these events. They're most likely few enough that they won't matter, and they should be small compared to the events that trigger them (that we've already allocated budget for).


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/rangefeed/event_size.go` line 91 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120001#-Nsn1G2t-PaPjjSj1fZv:-NsnDxT6-Nn-9dao1JJn:b-ykr0iu) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a428103984f0ff4b88637efed80ce1af83b624/pkg/kv/kvserver/rangefeed/event_size.go#L91)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Will these cause memory allocations? We should avoid those on the hot path (we do a bunch of them already, but should try not to add any more). You can find out by writing a small benchmark using `dev bench --bench-mem`, or using `scripts/goescape` to see if any of these escape to the heap.
</blockquote></details>

Good point. I think `ts.Get()` will always be the size of TimeStamp struct, but the size of span might need to be based on some evidence. I added a microbenchmark in this new PR https://github.com/cockroachdb/cockroach/pull/120347. I will compare it against the current calculateDataSize function later on. 


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
*[`pkg/kv/kvserver/rangefeed/event_size.go` line 110 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120001#-NsmzndPCajUO0jtogZu:-NsnEa-s438JjEVBW6H5:b7ubf0n) ([raw file](https://github.com/cockroachdb/cockroach/blob/67a428103984f0ff4b88637efed80ce1af83b624/pkg/kv/kvserver/rangefeed/event_size.go#L110)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

We shouldn't call `ForwardClosedTS` here, nor `rts.Init()` for `initRTS`. These will actually advance the rangefeed's resolved timestamp, which will cause incorrect behavior when a later call e.g. during `consumeEvent` sees `ForwardClosedTS` return `false` because the resolved timestamp was already advanced when computing the event size.
</blockquote></details>

Yeah : ( I realized this when I did more testing. I changed this to always publish checkpoint event in https://github.com/cockroachdb/cockroach/pull/120347. Since we are possibly over-accounting now, I introduced another PR https://github.com/cockroachdb/cockroach/pull/120348/commits/d303dae6ad8b1c53b49f4a16fc977dcd498057f7 to adjust the memory before publish to avoid holding on over-accounted memory from being held for too long. We can discuss more in our meeting. 


<!-- Sent from Reviewable.io -->
